### PR TITLE
173 add swaphelper test and improve tests

### PR DESCRIPTION
**Pull request by aliberkhsl**
#173

### CODE REVIEW

File: test/staking/SwapHelper.test.ts [hunk 1]
```diff
@@ -0,0 +1,499 @@
+import { ethers, upgrades } from "hardhat";
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:1]: 1. if you copy paste , copy from my latest tests
2. remove all unused variables and imports
3. all unrequire things in beforeAll

File: test/staking/SwapHelper.test.ts [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+    );
+    await genericCall(goodFundManager.address, encodedData, avatar, 0);
+  });
+
+  it("it should swap with multiple hops", async () => {
+    const goodFundManagerFactory = await ethers.getContractFactory(
+      "GoodFundManager"
+    );
+
+    const simpleStaking = await goodCompoundStakingFactory
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:377]: use the helper method from beforeAll

File: test/staking/SwapHelper.test.ts [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+    await setDAOAddress("DAI_ETH_ORACLE", daiEthOracle.address);
+    await setDAOAddress("MARKET_MAKER", marketMaker.address);
+  });
+
+  it("it should swap only safe amount when gains larger than safe amount", async () => {
+    const goodFundManagerFactory = await ethers.getContractFactory(
+      "GoodFundManager"
+    );
+
+    const simpleStaking = await goodCompoundStakingFactory
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:305]: use the helper method

File: test/staking/SwapHelper.test.ts [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+    const currentReserve = await pair.getReserves();
+
+    await simpleStaking.withdrawStake(ethers.utils.parseEther("100"), false);
+    expect(reserve[0].sub(currentReserve[0])).to.be.lt(currentGains[1]);
+
+    encodedData = goodFundManagerFactory.interface.encodeFunctionData(
+      "setStakingReward",
+      ["100", simpleStaking.address, currentBlock, currentBlock + 10000, true]
+    );
+    await genericCall(goodFundManager.address, encodedData, avatar, 0);
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:369]: 1. move all expects to the end, and make it clear what you are testing
2. it doesnt seem like your expects test what they should.
you want to test that the current gains are X and safe amount is Y, X>Y and collectInterest=Y
so there should be a test that after collection current gains are X-Y, or that the event has value of Y or that uniswap reserve has declined by Y
3. there seem to be lots of unrelevant code in this test
4. do the same for the next test.
5. as usuall any repeating code between tests that IS STILL RELEVANT, you should move to helpers

**Review by sirpy — CHANGES_REQUESTED** [review]

File: test/staking/SwapHelper.test.ts (hunk 3 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts]: Added helper for deploying staking

File: test/staking/SwapHelper.test.ts (hunk 1 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts:1]: Removed unusued variables and imports
Removed unrequired things from beforeAll

File: test/staking/SwapHelper.test.ts [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+    pair = new Contract(
+      pairAddress,
+      JSON.stringify(IUniswapV2Pair.abi),
+      staker
+    ).connect(founder);
+    setDAOAddress("CDAI", cDAI.address);
+    setDAOAddress("DAI", dai.address);
+
+    //This set addresses should be another function because when we put this initialization of addresses in initializer then nameservice is not ready yet so no proper addresses
+    await goodReserve.setAddresses();
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:149]: is this required?

File: test/staking/SwapHelper.test.ts [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+    );
+    await bat["mint(address,uint256)"](
+      founder.address,
+      ethers.utils.parseEther("5001000")
+    );
+    await usdc["mint(address,uint256)"](
+      founder.address,
+      ethers.utils.parseUnits("10000000", 6)
+    );
+    await bat.transfer(cBat.address, ethers.utils.parseEther("1000000")); // We should put extra BAT to mock cBAT contract in order to provide interest
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:177]: is this required?

File: test/staking/SwapHelper.test.ts [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+    await weth.deposit({ value: ethers.utils.parseEther("4000") });
+    console.log(
+      await weth.balanceOf(founder.address).then((_) => _.toString())
+    );
+    await weth.transfer(compPair.address, ethers.utils.parseEther("2000"));
+    await weth.transfer(daiPair.address, ethers.utils.parseEther("2000"));
+    console.log("minting liquidity pools");
+
+    await compPair.mint(founder.address);
+    await daiPair.mint(founder.address);
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:224]: all of these pairs and minting could probably go to a helper, DRY DRY DRY

File: test/staking/SwapHelper.test.ts [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+      "DaiEthPriceMockOracle"
+    );
+    daiEthOracle = await daiEthPriceMockFactory.deploy();
+
+    const ethUsdOracleFactory = await ethers.getContractFactory(
+      "EthUSDMockOracle"
+    );
+
+    batUsdOracle = await tokenUsdOracleFactory.deploy();
+    ethUsdOracle = await ethUsdOracleFactory.deploy();
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:164]: oracles also probably are used in other tests?
can it be reduced into a helper to be more DRY?

File: test/staking/SwapHelper.test.ts [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+      "BatUSDMockOracle"
+    );
+    setDAOAddress("UNISWAP_ROUTER", uniswapRouter.address);
+
+    daiUsdOracle = await tokenUsdOracleFactory.deploy();
+    const compUsdOracleFactory = await ethers.getContractFactory(
+      "CompUSDMockOracle"
+    );
+    compUsdOracle = await compUsdOracleFactory.deploy();
+    goodCompoundStaking = await deployStaking(
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:116]: aren't you creating a new goodcompount staking in each test?

File: test/staking/SwapHelper.test.ts [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+    await setDAOAddress("FUND_MANAGER", goodFundManager.address);
+    console.log("Deployed goodfund manager", {
+      manager: goodFundManager.address,
+    });
+
+    console.log("deployed contribution, deploying reserve...", {
+      founder: founder.address,
+    });
+
+    console.log("setting permissions...");
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:105]: not true

File: test/staking/SwapHelper.test.ts [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+      }
+    );
+    await setDAOAddress("FUND_MANAGER", goodFundManager.address);
+    console.log("Deployed goodfund manager", {
+      manager: goodFundManager.address,
+    });
+
+    console.log("deployed contribution, deploying reserve...", {
+      founder: founder.address,
+    });
```
Comment (sirpy) [inline test/staking/SwapHelper.test.ts:103]: not relevant

**Review by sirpy — CHANGES_REQUESTED** [review]

File: test/staking/SwapHelper.test.ts (hunk 5 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts:149]: not anymore ,removed

File: test/staking/SwapHelper.test.ts (hunk 6 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts:177]: Yeah for interest logic it's required that we should put extra token to ctoken in order to redeem tokens

File: test/staking/SwapHelper.test.ts (hunk 8 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts]: moved them to helper

File: test/staking/SwapHelper.test.ts (hunk 7 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts:202]: Moved comppair and daipair and related

File: test/staking/SwapHelper.test.ts (hunk 10 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts]: removed it

File: test/staking/SwapHelper.test.ts (hunk 11 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts]: removed

File: test/staking/SwapHelper.test.ts (hunk 9 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts]: yes removed it

File: test/staking/SwapHelper.test.ts (hunk 4 above)
Comment (aliberkhsl) [reply test/staking/SwapHelper.test.ts:222]: 1) Moved expects to the end
2) Improved expects
3)Removed non-relevant codes and moved others ones to helper as much as I could
4)Did
5)Did already as well

**Comment by sirpy** [comment]
@aliberkhsl 
what about simpleusdcstaking and usdcaavestaking

**Comment by sirpy** [comment]
@aliberkhsl did you test code coverage before and after removing all those tests?
make sure we didnt lose any coverage, then you can merge

**Pull request merged**